Wikipedia/UI-V5/WMFArticleHeaderImageGalleryViewController.m

```diff
@@ -95,45 +98,141 @@ - (UICollectionViewCell*)collectionView:(UICollectionView*)collectionView
         cell.imageView.contentMode = UIViewContentModeScaleAspectFill;
         return cell;
     }
+
     MWKImage* imageMetadata = self.images[indexPath.item];
+
+    // try to set cell's image from memory cache, otherwise fetch it and do face detection if needed
+    if (![self setCachedImageForCell:cell atIndexPath:indexPath metadata:imageMetadata]) {
+        [self fetchImage:imageMetadata forCellAtIndexPath:indexPath];
+    }
+
+    return cell;
+}
+
+- (NSInteger)collectionView:(UICollectionView*)collectionView numberOfItemsInSection:(NSInteger)section {
+    // if there are 0 images, show a placeholder
+    return self.images.count > 0 ? self.images.count : 1;
+}
+
+#pragma mark - Image Fetching & Setting
+
+/**
+ *  Fetch image data for the given `imageMetadata`, and insert it into the cell at the given `indexPath`.
+ *
+ *  @param imageMetadata    Metadata which specifies where to fetch the image and whether face detection needs to run.
+ *  @param indexPath        The `NSIndexPath` for the cell related to the image.
+ */
+- (void)fetchImage:(MWKImage*)imageMetadata forCellAtIndexPath:(NSIndexPath*)indexPath {
     @weakify(self);
     [[WMFImageController sharedInstance] fetchImageWithURL:[imageMetadata sourceURL]]
-    .then(^id (UIImage* image) {
+    .then(^id (WMFImageDownload* download) {
+        @strongify(self);
+        UIImage* image = download.image;
         if (!self) {
             return [NSError cancelledError];
+        }
+        BOOL shouldAnimate = ![download.origin isEqualToString:[WMFImageDownload imageOriginMemory]];
```

Comment:
this is the lowest-common-scope where it's used. Would prefer not to re-check in both places, DRY and all that.